### ISSUE TITLE
Check signature of `__init__` if comparing against a `Protocol`.

### ISSUE THREAD

**Issue by randolf-scholz** [microsoft/pyright#7249]
Both `mypy` and `pyright` do not validate subtyping for the `__init__` method, since it would generally introduce a lot of noise.

However, sometimes I wish it were possible to opt-in to `__init__` checking in some circumstances, so I was wondering if it would be possible/sensible to make an exemption to this rule when dealing with `Protocol`-types, if the `Protocol` specifically describes a `__init__` signature.

One use case I have, from a machine learning context, is that quite often some model class needs to initialize several subcomponents. Such a subcomponent can often be one of several options, and so we typically need all the options to share compatible constructor methods. It would be nice to be able to type hint this properly.

An example:

```python
class Cell(Protocol):
    input_size: int
    hidden_size: int

    @abstractmethod
    def __init__(self, /, input_size: int, hidden_size: int) -> None: ...
    @abstractmethod
    def __call__(self, y: Tensor, x: Tensor, /) -> Tensor: ...


AVAILABLE_CELLS: dict[str, type[Cell]] = {
    "RNNCell": torch.nn.RNNCell,  # <- validate __init__ compatibility!
    "GRUCell": torch.nn.GRUCell,
    "LSTMCell": torch.nn.LSTMCell,
}


class RecurrentNetwork(nn.Module):
    input_size: int
    hidden_size: int
    cell: Cell  # actually want intersection nn.Module & Cell

    def __init__(
        self,
        input_size: int,
        hidden_size: int,
        cell_type: str | type[Cell] = "RNNCell",
    ) -> None:
        super().__init__()
        cell_class = (
            AVAILABLE_CELLS[cell_type] if isinstance(cell_type, str) else cell_type
        )
        self.cell = cell_class(input_size, hidden_size)
        ...


model = RecurrentNetwork(5, 5, CustomCell)  # <- validate __init__ compatibility!
```

**Comment by erictraut** [comment]
I've been working to define intended behaviors (eliminate ambiguities and under-specified areas) within the typing spec and then make pyright [conformant](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) with the updated spec. 

The behavior you're discussing here falls within an under-specified part of the spec. It's on a long list of items that I'd like to eventually address. In the meantime, I don't plan to make any changes in this area because I don't want to add functionality that I then need to reverse later because of a spec update. For that reason, I'm going to close this issue for now. I'm open to revisiting it at a later time.
